src/marqo/core/document/add_documents_handler.py

```diff
@@ -0,0 +1,562 @@
+
+
+class AddDocumentsHandler(ABC):
+
+    def __init__(self, marqo_index: MarqoIndex, config: Config, add_docs_params: AddDocsParams):
+        self.marqo_index = marqo_index
+        self.add_docs_params = add_docs_params
+        self.config = config
+        self.add_docs_response_collector = AddDocumentsResponseCollector()
+        self.tensor_fields_container = self._create_tensor_fields_container()
+
+    def add_documents(self):
+        """
+        Template method for adding documents to Marqo index
+        """
+        with RequestMetricsStore.for_request().time("add_documents.processing_before_vespa"):
+            for loc, original_doc in enumerate(reversed(self.add_docs_params.docs)):
+                doc = copy.deepcopy(original_doc)
+                original_id = None
+                try:
+                    self.validate_doc(doc)
+
+                    original_id = self.validate_and_pop_doc_id(doc)
+                    doc_id = original_id or str(uuid.uuid4())
+                    marqo_doc = {ORIGINAL_ID: original_id, MARQO_DOC_ID: doc_id}  # keep this info for error report
+
+                    for field_name, field_content in doc.items():
+                        self.handle_field(marqo_doc, field_name, field_content)
+
+                    self.handle_multi_modal_fields(marqo_doc)
+
+                    self.add_docs_response_collector.collect_marqo_doc(loc, marqo_doc)
+                except AddDocumentsError as err:
+                    self.add_docs_response_collector.collect_error_response(original_id, err, loc)
+
+            # retrieve existing docs for existing tensor
+            if self.add_docs_params.use_existing_tensors:
+                # TODO capture the telemetry data for retrieving exiting docs?
+                result = self.config.vespa_client.get_batch(list(self.add_docs_response_collector.visited_doc_ids),
+                                                            self.marqo_index.schema_name)
+                existing_vespa_docs = [r.document for r in result.responses if r.status == 200]
+                self.handle_existing_tensors(existing_vespa_docs)
+
+            # vectorise tensor fields
+            self.vectorise_tensor_fields()
+
+        # FIXME this step is not timed in the original implementation
+        vespa_docs = self.convert_to_vespa_docs()
+
+        self.pre_persist_to_vespa()
+
+        # persist to vespa if there are still valid docs
+        with RequestMetricsStore.for_request().time("add_documents.vespa._bulk"):
+            response = self.config.vespa_client.feed_batch(vespa_docs, self.marqo_index.schema_name)
+
+        with RequestMetricsStore.for_request().time("add_documents.postprocess"):
+            self.handle_vespa_response(response)
+            return self.add_docs_response_collector.to_add_doc_responses(self.marqo_index.name)
+
+    @abstractmethod
+    def _create_tensor_fields_container(self) -> TensorFieldsContainer:
+        pass
+
+    @abstractmethod
```

Comment:
Can we have docstrings for these methods so it's clear what the contract is and what the abstraction means?

Reply:
Will do